File: python3/leetcodepy/surrounded_regions.py
"""
Given an m x n matrix board containing 'X' and 'O', capture all regions that are 4-directionally surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.



Example 1:
https://assets.leetcode.com/uploads/2021/02/19/xogrid.jpg

Input: board = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
Output: [["X","X","X","X"],["X","X","X","X"],["X","X","X","X"],["X","O","X","X"]]
Explanation: Surrounded regions should not be on the border, which means that any 'O' on the border of the board are not flipped to 'X'. Any 'O' that is not on the border and it is not connected to an 'O' on the border will be flipped to 'X'. Two cells are connected if they are adjacent cells connected horizontally or vertically.

Example 2:

Input: board = [["X"]]
Output: [["X"]]


Constraints:

m == board.length
n == board[i].length
1 <= m, n <= 200
board[i][j] is 'X' or 'O'.
"""
from typing import List


class Solution1:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n = len(board), len(board[0])
        for i in range(m):
            self._mark(board, m, n, i, 0)
            if n != 1:
                self._mark(board, m, n, i, n - 1)
        for j in range(n):
            self._mark(board, m, n, 0, j)
            if m != 1:
                self._mark(board, m, n, m - 1, j)

        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                elif board[i][j] == 'M':
                    board[i][j] = 'O'

    def _mark(self, board: List[List[str]], m: int, n: int, i: int, j: int):
        if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != 'O':
            return
        board[i][j] = 'M'
        self._mark(board, m, n, i - 1, j)
        self._mark(board, m, n, i, j - 1)
        self._mark(board, m, n, i + 1, j)
        self._mark(board, m, n, i, j + 1)

File: java/src/main/java/io/binac/leetcode/SurroundedRegions.java
package io.binac.leetcode;

/**
 * <p>Given an <code>m x n</code> matrix <code>board</code> containing <code>'X'</code> and <code>'O'</code>, <em>capture all regions that are 4-directionally&nbsp;surrounded by</em> <code>'X'</code>.</p>
 *
 * <p>A region is <strong>captured</strong> by flipping all <code>'O'</code>s into <code>'X'</code>s in that surrounded region.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2021/02/19/xogrid.jpg" style="width: 550px; height: 237px;">
 * <pre><strong>Input:</strong> board = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
 * <strong>Output:</strong> [["X","X","X","X"],["X","X","X","X"],["X","X","X","X"],["X","O","X","X"]]
 * <strong>Explanation:</strong> Surrounded regions should not be on the border, which means that any 'O' on the border of the board are not flipped to 'X'. Any 'O' that is not on the border and it is not connected to an 'O' on the border will be flipped to 'X'. Two cells are connected if they are adjacent cells connected horizontally or vertically.
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> board = [["X"]]
 * <strong>Output:</strong> [["X"]]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>m == board.length</code></li>
 * 	<li><code>n == board[i].length</code></li>
 * 	<li><code>1 &lt;= m, n &lt;= 200</code></li>
 * 	<li><code>board[i][j]</code> is <code>'X'</code> or <code>'O'</code>.</li>
 * </ul>
 */
public class SurroundedRegions {
    public static class Solution1 {
        private void mark(char[][] board, int i, int j, int m, int n) {
            if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != 'O')
                return;
            board[i][j] = 'M';
            mark(board, i - 1, j, m, n);
            mark(board, i + 1, j, m, n);
            mark(board, i, j - 1, m, n);
            mark(board, i, j + 1, m, n);
        }

        public void solve(char[][] board) {
            final int m = board.length, n = board[0].length;

            for (int i = 0; i < m; ++i) {
                mark(board, i, 0, m, n);
                mark(board, i, n - 1, m, n);
            }
            for (int j = 0; j < n; ++j) {
                mark(board, 0, j, m, n);
                mark(board, m - 1, j, m, n);
            }

            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    switch (board[i][j]) {
                        case 'O':
                            board[i][j] = 'X';
                            break;
                        case 'M':
                            board[i][j] = 'O';
                            break;
                    }
                }
            }
        }
    }
}
